include/RAJA/policy/cuda/kernel/CudaKernel.hpp

```diff
@@ -182,77 +220,126 @@ struct CudaLaunchHelper;
- * determine the number of threads and blocks
+ * Helper class specialization to determine the number of threads and blocks.
+ * The user may specify the number of threads and blocks or let one or both be
+ * determined at runtime using the CUDA occupancy calculator.
  */
-template<int num_threads, bool async0, typename StmtList, typename Data>
-struct CudaLaunchHelper<cuda_occ_calc_launch<num_threads, async0>,StmtList,Data>
+template<bool async0, int num_blocks, int num_threads, typename StmtList, typename Data>
+struct CudaLaunchHelper<cuda_launch<async0, num_blocks, num_threads>,StmtList,Data>
 {
+  using Self = CudaLaunchHelper;
+
   static constexpr bool async = async0;
 
   using executor_t = internal::cuda_statement_list_executor_t<StmtList, Data>;
 
-  inline static void max_blocks_threads(int shmem_size,
-      int &max_blocks, int &max_threads)
+  inline static void recommended_blocks_threads(int shmem_size,
+      int &recommended_blocks, int &recommended_threads)
   {
-
     auto func = internal::CudaKernelLauncher<Data, executor_t>;
 
-    cudaOccupancyMaxPotentialBlockSize(&max_blocks,
-                                       &max_threads,
-                                       func,
-                                       shmem_size);
+    if (num_blocks <= 0) {
 
-  }
+      if (num_threads <= 0) {
 
-  static void launch(Data const &data,
-                     internal::LaunchDims launch_dims,
-                     size_t shmem,
-                     cudaStream_t stream)
-  {
+        //
+        // determine blocks at runtime
+        // determine threads at runtime
+        //
+        internal::cuda_occupancy_max_blocks_threads<Self>(
+            func, shmem_size, recommended_blocks, recommended_threads);
 
-    auto func = internal::CudaKernelLauncher<Data, executor_t>;
+      } else {
 
-    func<<<launch_dims.blocks, launch_dims.threads, shmem, stream>>>(data);
-  }
-};
+        //
+        // determine blocks at runtime
+        // threads determined at compile-time
+        //
+        recommended_threads = num_threads;
 
+        internal::cuda_occupancy_max_blocks<Self, num_threads>(
+            func, shmem_size, recommended_blocks);
 
+      }
 
-/*!
- * Helper class specialization to use the CUDA device properties and a user
- * specified number of threads to compute the number of blocks/threads
- */
-template<bool async0, int num_blocks, int num_threads, typename StmtList, typename Data>
-struct CudaLaunchHelper<cuda_explicit_launch<async0, num_blocks, num_threads>,StmtList,Data>
-{
-  static constexpr bool async = async0;
+    } else {
 
-  using executor_t = internal::cuda_statement_list_executor_t<StmtList, Data>;
+      if (num_threads <= 0) {
+
+        //
+        // determine threads at runtime, unsure what use 1024
+        // this value may be invalid for kernels with high register pressure
+        //
+        recommended_threads = 1024;
```

Comment:
It's set here and below in max_threads

Reply:
Yeah, I saw that after I wrote my comment. I should have resolved it. Thanks!